### PR TITLE
Independent PID parameters for each extruder

### PR DESCRIPTION
**Pull request by grob6000**
tl;dr: New feature - each extruder gets it's own set of PID parameters

Background: Singular PID parameters for the extruders presents problems when using a spare extruder channel for a chamber heater, or where two extruders have very different construction (e.g. pairing E3D and E3D volcano, this has been suggested on soliforum.com). Implementing chamber heater commands would require support from hosts.

Fix:
* Variables Kp, Ki, Kd, Kc are now arrays of length EXTRUDERS
* EEPROM saves all values, space allocated for up to 3 extruders (as per marlin current max) - new version
* M301 takes E parameter (optional, default=0) to set each extruder (M301 [E?] [P?] [I?] [D?] [C?]). Report now "ok e:0 p:22 i:2 d:60 c:1" - should be confirmed to be compatible with any hosts that use this (in my experience usually only done via terminal)
* EEPROM output unchanged for compatibility with hosts (i.e. Repetier), although this hides that multiple PID values are being saved...

Open to suggestions, comments, thank you for your time.

### CODE REVIEW

**Comment by boelle** [comment]
build failed.  3e2af67

the fix R-H thing

**Comment by bluesign2k** [comment]
I like this idea and think it would be useful. There have been a number of times where I've had to compromise with PID values to get get two dissimilar heads working.
I think if nothing else it would be good to set some initial secondary PID values in configuration.h.
I'm not sure why @boelle has closed this already - just because some code might need some further modification before it's merged doesn't mean the pull needs to be closed! Frankly, in this case if it breaks host compatability to add useful features then it's up to the host developers to fix. Particularly in this case where their fix would 've easy... isn't that also part the point of having a development branch?

**Comment by boelle** [comment]
i closed it as it did not pass build checks

it needs to be checked and resubmitted

i can reopen but it cant be merged before it passes

Den lørdag den 10. januar 2015 skrev Chris <notifications@github.com>:

**Comment by grob6000** [comment]
Aha, looks like I missed use of K parameters in the LCD files - oops...

**Comment by boelle** [comment]
hehe... lets see if it passes when you fik that one...

**Comment by grob6000** [comment]
Looks like that did it. May as well let the ultralcd edit all the PID values, what could possibly go wrong haha. Recommend someone with an LCD panel tests this first though... I can always pare it back to the LCD edits PID values for E0 only if preferred.

**Comment by boelle** [comment]
yep, just having lunch... @ErikZalm & @stv0g  any comments on a merge?

**Comment by grob6000** [comment]
Messy though it would be, should I `#ifdef` this as an option for now? I'm worried it may have undesired consequences for multiple-extruder users, who may not realise that they have to update the PID parameters for each extruder separately. In this (probably more common) use case, it's actually preferable to have a single set of parameters...

**Comment by boelle** [comment]
multple sets of pid values ok, as long there is a comment on it in the config so people know it...

**Comment by boelle** [comment]
@alexborro also any comment on this one? should we clear bugs before we merge?

**Comment by ErikZalm** [comment]
@boelle If users use the "old" config file there should be no difference.
A comment is not enough. Many users do not check the config file.

**Comment by boelle** [comment]
would they look at the front page with all the features? or will there be a better place to put the info ?

**Comment by grob6000** [comment]
I put the switch (commented out) into the default configuration.h (see PID_PARAMS_PER_EXTRUDER), if the old config file is used it should be undef and default to single-parameters as before - no functional change, even M301 will work the same.

**Comment by boelle** [comment]
i thought of adding new features to the release list... currently i only have listed what we fix

**Comment by grob6000** [comment]
This is definitely new-feature territory, I'm aware of the bugathon. I'm not fussed if you want to hold off. I'm happy to merge and patch later if there's a lot of change. :)

**Comment by boelle** [comment]
hehe... me just new still to doing merges and dont want to break anything... 

but think its time to add new features added to the release list.... might do that during the day...

@grob6000 are you good at writing code? maybe you could help us with the bugs? no pressure here

**Comment by grob6000** [comment]
I'd love to help, I'm a little new to arduino and github (remember the accidental pull?) but have done a bunch of native AVR C before it was cool, so it's all pretty familiar. Not sure whether that counts as 'good' though haha. I'll have a look at the bugs, if there's anything I can replicate on my machine I'll pipe up (in the relevant thread) and have a go.

**Comment by boelle** [comment]
we all start from somewhere and nobody is complete...

just do the best you can....

as for the bugs... if you submit a pull request that fixes a bug please mention what bug each PR solves :-D makes it faster to close the bug

**Comment by alexborro** [comment]
I think this one should be merged.
Em 11/01/2015 10:23, "Bo Herrmannsen" <notifications@github.com> escreveu:

**Comment by boelle** [comment]
ok, will do  so if no other protests during the day

**Comment by ErikZalm** [comment]
Only merge it if older config file work without a problem.
And maybe we should improve the eprom structure. Now people need to reprogram their eeprom part because we added some variables.

**Comment by boelle** [comment]
was thinking of adding to info on how to use this on the documentation... but where to put it?

**Comment by boelle** [comment]
oh yes... 3 hours ago you said that it should not be a problem:

"@boelle If users use the "old" config file there should be no difference."

**Comment by grob6000** [comment]
Yup, probably needs a few eyes on, particularly the LCD changes.

EEPROM is a whole other issue... I was looking into fixed addresses in the h file (eliminates the need for ordered and dummy r/w's) and making sure all the unused parameters are written as a sane default value... Currently some quite reasonable config changes (add an lcd for example) can change the structure of the eeprom without changing the version.

**Comment by boelle** [comment]
but where could i add documentation?

**Comment by alexborro** [comment]
I think we need someone to start Marlin Documentation. So far most of
features description and usage instruction is inside configuration.h.

Cheers.

Alex
Em 11/01/2015 13:12, "Bo Herrmannsen" <notifications@github.com> escreveu:

**Comment by scotty1024** [comment]
The coin of the editable documentation realm is a docx file. The coin of the reading realm is a PDF file.

I suggest creating a docx file AKA XML Word file. As it is updated it is "compiled" into PDF. Both files are then checked into the Marlin github repo.

There are many places to update docx files online using web tools. And many open source tools. There are various paid and unpaid apps for phones and tablets. And of course you can use Microsoft Word itself.

**Comment by alexborro** [comment]
What about something like a Wiki? Easy to use, powerful and collaborative,
I mean, you can control write permission. And no special tool/app needed.
Em 11/01/2015 15:52, "scotty1024" <notifications@github.com> escreveu:

**Comment by nophead** [comment]
You get one automatically with github so Marlin already has one: https://github.com/MarlinFirmware/Marlin/wiki

**Comment by scotty1024** [comment]
How do you tag the Wiki with a release branch?

**Comment by nophead** [comment]
I don't think you can. That is the problem with wikis and the reason I don't use them.

**Comment by boelle** [comment]
@other collabs: should we merge this one?

**Comment by alexborro** [comment]
I think so.
Em 13/01/2015 08:28, "Bo Herrmannsen" notifications@github.com escreveu:

**Comment by boelle** [comment]
oki... we cant sit on our fingers either :-P

**Comment by MagoKimbra** [comment]
One question??
Kc isn't used... I search in fw, but i do not see this value....

**Comment by grob6000** [comment]
Kc seemed to be used in the PID, the switch was on in the default config by default so I made sure I took it seriously haha.

Note that to do the switch I use a macro (PID_PARAM(param, e)), so other than in configurationstore.h Kc may not appear in code searches... If I have mucked it up please let me know, I'll have a look when I get a chance tonight. :)

**Comment by grob6000** [comment]
AFAIK, Kc is intended to increase the heating power slightly in concert with the extrusion speed, to compensate for the heat being taken out by the molten filament.

Digging into the temp routine a bit, it looks like Kc is actually not implemented in the PID routine (duh! silly me not spotting that), but it's assignment & config switch must have been left in. I don't know whether it's widely used, I'd think the best way to deal with that would be to remove it from Marlin altogether. Certainly excellent printing results at a large range of speeds seem to be achieved without it.

Does anyone know the history of Kc?

**Comment by thinkyhead** [comment]
Kc isn't mentioned in the RepRap wiki, but it does show up in this [RepRap forum topic](http://forums.reprap.org/read.php?147,364712) which includes the code line

```
cTerm[e] = (lpq[lpq_ptr] / axis_steps_per_unit[E_AXIS]) * Kc;
```
